File: app/javascript/channels/conversation_channel.js
import consumer from "./consumer"

document.addEventListener('turbo:load', () => {
    const messagesContainer = document.getElementById('messages')

    if (messagesContainer) {
        const conversationId = messagesContainer.dataset.conversationId

        consumer.subscriptions.create({ channel: "ConversationChannel", conversation_id: conversationId }, {
            connected() {
                // Called when the subscription is ready for use on the server
                console.log("Connected to conversation channel")
            },

            disconnected() {
                // Called when the subscription has been terminated by the server
            },

            received(data) {
                // Called when there's incoming data on the websocket for this channel
                if (data.conversation_id === parseInt(conversationId)) {
                    const messageElement = document.createElement('div')
                    messageElement.innerHTML = data.message
                    messagesContainer.appendChild(messageElement.firstChild)
                    messagesContainer.scrollTop = messagesContainer.scrollHeight
                }
            }
        })

        messagesContainer.scrollTop = messagesContainer.scrollHeight
    }
})

File: app/channels/conversation_channel.rb
class ConversationChannel < ApplicationCable::Channel
  def subscribed
    stream_from "conversation_channel_#{params[:conversation_id]}"
  end

  def unsubscribed
    stop_all_streams
  end

  def send_message(data)
    Message.create!(
      conversation_id: params[:conversation_id],
      user_id: params[:user_id],
      content: data["message"],
      message_type: data["message_type"],
    )
  end
end